src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs

```diff
@@ -865,6 +876,70 @@ async Task<IEnumerable<IAccount>> ITokenCacheInternal.GetAccountsAsync(Authentic
+                return null;
+            }
+
+            IEnumerable<MsalAccountCacheItem> accountCacheItems = _accessor.GetAllAccounts();
+
+            // Multi-cloud support - must filter by env.
+            ISet<string> allEnvironmentsInCache = new HashSet<string>(
+                accountCacheItems.Select(aci => aci.Environment),
+                StringComparer.OrdinalIgnoreCase);
+
+            AdalUsersForMsal adalUsersResult = null;
+
+            if (ServiceBundle.Config.LegacyCacheCompatibilityEnabled)
+            {
+                adalUsersResult = CacheFallbackOperations.GetAllAdalUsersForMsal(
+                    Logger,
+                    LegacyCachePersistence,
+                    ClientId);
+                allEnvironmentsInCache.UnionWith(adalUsersResult.GetAdalUserEnviroments());
+            }
+
+            InstanceDiscoveryMetadataEntry instanceMetadata = await ServiceBundle.InstanceDiscoveryManager.GetMetadataEntryTryAvoidNetworkAsync(
+                requestParameters.AuthorityInfo,
+                allEnvironmentsInCache,
+                requestParameters.RequestContext).ConfigureAwait(false);
+
+            return GetTenantProfiles(instanceMetadata, homeAccountId);
+        }
+
+        private IDictionary<string, ITenantProfile> GetTenantProfiles(InstanceDiscoveryMetadataEntry instanceMetadata, string homeAccountId)
+        {
+            if (string.IsNullOrEmpty(homeAccountId) || string.IsNullOrEmpty(AccountId.ParseFromString(homeAccountId).TenantId))
+            {
+                return null;
+            }
+
+            IDictionary<string, ITenantProfile> tenantProfiles = new Dictionary<string, ITenantProfile>();
+            bool filterByClientId = !_featureFlags.IsFociEnabled;
+
+            // Get id tokens in the cache and filter them with environment and home account id to get the id tokens for the account.
+            IEnumerable<MsalIdTokenCacheItem> idTokenCacheItems = GetAllIdTokensWithNoLocks(filterByClientId);
+            idTokenCacheItems = idTokenCacheItems.Where(idToken => instanceMetadata.Aliases.ContainsOrdinalIgnoreCase(idToken.Environment));
+            idTokenCacheItems = idTokenCacheItems.Where(idToken => homeAccountId.Equals(idToken.HomeAccountId));
+
+            foreach (MsalIdTokenCacheItem idTokenCacheItem in idTokenCacheItems.ToList())
+            {
+                tenantProfiles[idTokenCacheItem.TenantId] = GetTenantProfile(idTokenCacheItem);
+            }
+
+            return tenantProfiles;
+        }
+
+        private ITenantProfile GetTenantProfile(MsalIdTokenCacheItem msalIdTokenCacheItem)
+        {
+            AccountId homeAccountId = AccountId.ParseFromString(msalIdTokenCacheItem.HomeAccountId);
+            bool isHomeTenant = homeAccountId.TenantId.Equals(msalIdTokenCacheItem.TenantId);
+            
+            return new TenantProfile(msalIdTokenCacheItem.IdToken.ObjectId, msalIdTokenCacheItem.TenantId, msalIdTokenCacheItem.Secret, isHomeTenant);
```

Comment:
null check here, and can you use IdToken instead of Secret?

Reply:
No secret is the idtoken in jwt format which contains all the claims. IdToken is the class which parses the secret and stores some of the claims.

Reply:
ok. thanks.